### ISSUE TITLE
reload(numpy);numpy.max([]) raises TypeError

### ISSUE THREAD

**Issue by cgohlke** [numpy/numpy#7844]
Using `maintenance/1.11.x` on Windows (msvc builds), the following code fails:
```Python
Python 3.5.2 (v3.5.2:4def2a2901a5, Jun 25 2016, 22:18:55) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> from importlib import reload
>>> import numpy
>>> reload(numpy)
<module 'numpy' from 'X:\\Python35\\lib\\site-packages\\numpy\\__init__.py'>
>>> numpy.max([])
Traceback (most recent call last):
  File "X:\Python35\lib\site-packages\numpy\core\fromnumeric.py", line 2290, in amax
    amax = a.max
AttributeError: 'list' object has no attribute 'max'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "X:\Python35\lib\site-packages\numpy\core\fromnumeric.py", line 2293, in amax
    out=out, **kwargs)
  File "X:\Python35\lib\site-packages\numpy\core\_methods.py", line 26, in _amax
    return umr_maximum(a, axis, None, out, keepdims)
TypeError: an integer is required (got type type)
```
I traced this to #7736, backport of #4619

**Comment by pv** [comment]
The _NoValue sentinel changes on reload, but not its values captured in 
function default arguments. The problem is referring to it via `import 
numpy as np; np._NoValue` instead of `from numpy import _NoValue` in 
fromnumeric.py, which is not reload-safe.

**Comment by charris** [comment]
@pv Good catch.

**Comment by charris** [comment]
I wonder if it would make more sense for  `_NoValue` to be locally defined in the modules? It is a private class in any case.

**Comment by charris** [comment]
That said, I thought C modules could not be reloaded?

**Comment by charris** [comment]
Curiously, there is no error with 2.7 or 3.5 in Fedora 24.

**Comment by pv** [comment]
Yes, I think the reload-safe solution is to use module-specific private 
sentinels. Otherwise, different modules end up with different singletons 
on reload. Or maybe just put the sentinel in `numpy.core` etc. private 
place, which people are unlikely to reload.

**Comment by cgohlke** [comment]
I can confirm that using `from numpy import _NoValue` instead of `import numpy as np; np._NoValue` fixes this issue.

**Comment by charris** [comment]
Looks like we need a global, reload safe, `_NoValue`, as calls from other modules pass through `fromnumeric` which checks the values. The quick work around is probably to define it somewhere more reload safe than `numpy/__init__.py`.

**Comment by njsmith** [comment]
Possible alternative approach: add something like this to `numpy/__init__.py`:
```python
if "_already_loaded" in globals():
    raise RuntimeError("Reloading numpy isn't supported")
_already_loaded = True
```

**Comment by charris** [comment]
Probably the explicit import works also, looks like the local binding hangs onto the instance in the original import of Numpy before any reloads.

This all makes me wonder if the various warnings defined in `__init__.py` are also safe if used in `catch_warnings`. We don't do that, just wondering.

**Comment by charris** [comment]
@njsmith Yeah, reloading numpy alone doesn't look to achieve much except mess up the local stuff defined in the `__init__.py`.

**Comment by charris** [comment]
I note that `isinstance` also fails for reloaded classes. I'm inclined to follow @njsmith and just disallow reloading numpy, it isn't clear to me that reloading does anything useful. Thoughts?

**Comment by charris** [comment]
Proposed fix in #7853.
